### PR TITLE
add snapshot functionality

### PR DESCRIPTION
**Pull request by mfussenegger**

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -5,6 +5,10 @@ Changes for Crate
 Unreleased
 ==========
 
+ - Added the ``CREATE REPOSITORY``, ``DROP REPOSITORY``, ``CREATE SNAPSHOT``,
+   ``RESTORE SNAPSHOT`` statements and the ``sys.snapshots`` and
+   ``sys.repositories`` virtual tables for backup and restore functionality.
+
```
Comment (dobe) [inline CHANGES.txt]: i think its enough to say that the snapshot functionality for backup and restore is implemented.  since the documentation has more information

File: app/src/main/dist/config/crate.yml [hunk 2]
```diff
@@ -233,10 +233,16 @@
 #
 # path.plugins: /path/to/plugins
 
+
+# Paths where Repositories of type fs may be created:
+#
+# path.repo: /path/to/shared/fs,/other/shared/fs
```
Comment (dobe) [inline app/src/main/dist/config/crate.yml]: should probably go to the backup /restore section?

File: docs/configuration.txt [hunk 3]
```diff
@@ -534,6 +549,43 @@ Blobs
   By default blobs will be stored under the same path as normal data.
   A relative path value is interpreted as relative to ``CRATE_HOME``.
 
+.. _ref-configuration-repositories:
+
+Repositories
+------------
```
Comment (dobe) [inline docs/configuration.txt]: it is not clear that this section is for URL type, also the above section with path.repo - it is not clear that this is for fs type. so probably it would make sense to structure this document diferently e.g. by repo type

File: docs/sql/reference/create_snapshot.txt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    [ WITH (snapshot_parameter [= value], [, ...]) ]
+
+The following configuration parameters can be used to modify how the snapshot is
+created:
+
+:wait_for_completion: (Default ``false``) By default the request returns once the snapshot creation
+                      started. If set to ``true`` the request returns after the
+                      whole snapshot was created or an error occurred.
+                      This can take a serious amount of time.
```
Comment (dobe) [inline docs/sql/reference/create_snapshot.txt]: point to the documentation about sys.snapshots to show the alternative way of tracking snaphsot state

File: docs/sql/reference/drop_snapshot.txt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    DROP SNAPSHOT repository_name.snapshot_name
+
+
+Description
+===========
+
+Delete a snapshot from a repository and all files only referenced by this snapshot.
+
+If this statement is executed against a snapshot that is currently being created,
+the creation is aborted and all files created so far get deleted.
```
Comment (dobe) [inline docs/sql/reference/drop_snapshot.txt]: are deleted

File: docs/sql/reference/restore_snapshot.txt [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+Restore one or more tables or partitions from an existing snapshot into the cluster.
+The snapshot must be given as fully qualified reference with ``repository_name``
+and ``snapshot_name``.
+
+It is possible to restore all tables contained in the snapshot using the ``ALL``
+keyword. Single tables and/or partitions can be selected for restoring by giving them
+as ``table_ident`` or partition reference given the ``partition_column`` values.
+
+It is possible to change the table names that get restored and to change the
+table settings of the restored table.
```
Comment (dobe) [inline docs/sql/reference/restore_snapshot.txt]: this is NOT implemented

File: docs/sql/reference/restore_snapshot.txt [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+restored to the cluster:
+
+:ignore_unavailable: (Default ``false``) Per default the restore command fails if a table
+                     is given that does not exist in the snapshot. If set to ``true`` those
+                     missing tables are ignored.
+
+:wait_for_completion: (Default: ``false``) By default the request returns once the restore
+                      operation started. If set to ``true`` the request returns after all
+                      selected tables from the snapshot are restored or an error occurred.
+                      This can take a serious amount of time.
```
Comment (dobe) [inline docs/sql/reference/restore_snapshot.txt]: same here. link to alternative

File: docs/sql/snapshot_restore.txt [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+CRATE supports different repository types:
+:ref:`ref-create-repository-types-fs`, :ref:`ref-create-repository-types-hdfs`,
+and :ref:`ref-create-repository-types-url`. Support for further types
+might be added using :ref:`plugins`.
+
+The creation of a repository simply configures it inside the crate
+cluster. In general no data is written, no snapshots inside
+repositories are changed or deleted. This way you can tell the CRATE
+cluster about repositories that already contains snapshots.
```
Comment (dobe) [inline docs/sql/snapshot_restore.txt]: about existing repositories which already contain snapshots

File: docs/sql/snapshot_restore.txt [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+A snapshot is referenced by the name of the repository and the
+snapshot name, separated by a dot. If used with ``ALL``, all tables of
+the cluster (except :ref:`sys-tables <system-information>`,
+:ref:`information_schema` tables and blob tables) along with all
+persistent settings and the full cluster metadata are stored inside
+the snapshot.
+
+It is possible to select any number of tables for inclusion in the
+snapshot. In this case only the necessary metadata for those tables is
+included in the snapshot and the persistent settings will be excluded.
```
Comment (dobe) [inline docs/sql/snapshot_restore.txt]: is this metadata thingy relevant to the user? what are persistent settings in this regard?

File: docs/sql/snapshot_restore.txt [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+
+.. note::
+
+   After installing all needed libraries, every node must be restarted.
+
+Hadoop 1.x required libraries
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When using `Hadoop (1.x)`_ version 1.x, please install only following
+libraries into Crate's CLASSPATH:
```
Comment (dobe) [inline docs/sql/snapshot_restore.txt]: the "only" is not required anymore

File: docs/sql/snapshot_restore.txt [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+.. note::
+
+   As of writing this documentation, the latest stable `Hadoop 1.x`
+   version is **1.2.1**. Required libraries may differ for other
+   versions.
+
+Hadoop 2.x required libraries
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When using `Hadoop (YARN)`_ version 2.x, please install only following
```
Comment (dobe) [inline docs/sql/snapshot_restore.txt]: same here for "only"

File: docs/sql/snapshot_restore.txt [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+Restore
+-------
+
+Once a snapshot is created, it can be used to restore its tables to
+the state when the snapshot was created.
+
+To get basic information about snapshots the :ref:`sys.snapshots
+<sys-snapshots>` table can be queried::
+
+    cr> SELECT "repository", name, state, concrete_indices
```
Comment (dobe) [inline docs/sql/snapshot_restore.txt]: do we need to reserve this name? if not add to non-reserved and remove quotes. also for snapshot

**Comment by dobe** [comment]
missing documentation for canceling restore https://www.elastic.co/guide/en/elasticsearch/guide/current/_restoring_from_a_snapshot.html#_canceling_a_restore

**Comment by dobe** [comment]
missing feature monitoring restore https://www.elastic.co/guide/en/elasticsearch/guide/current/_restoring_from_a_snapshot.html#_monitoring_restore_operations

**Comment by bogensberger** [comment]
ack
